runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java

```diff
@@ -242,10 +327,41 @@ public String toString() {
   }
 
   /** Advance the WMs onBatchCompleted event. */
-  public static class WatermarksListener extends JavaStreamingListener {
+  public static class WatermarkAdvancingStreamingListener extends JavaStreamingListener {
+    private static final Logger LOG =
+        LoggerFactory.getLogger(WatermarkAdvancingStreamingListener.class);
+
+    private long timeOf(JavaBatchInfo info) {
+      return info.batchTime().milliseconds();
+    }
+
+    private long laterOf(long t1, long t2) {
+      return Math.max(t1, t2);
+    }
+
     @Override
     public void onBatchCompleted(JavaStreamingListenerBatchCompleted batchCompleted) {
-      GlobalWatermarkHolder.advance();
+
+      final long currentBatchTime = timeOf(batchCompleted.batchInfo());
+
+      GlobalWatermarkHolder.advance(Long.toString(currentBatchTime));
+
+      // make sure to update the last watermarked batch time AFTER the watermarks have already
```

Comment:
The watermarks block is updated asynchronously, perhaps we should include this fact in this comment?